web3/middleware/stalecheck.py

```diff
@@ -0,0 +1,55 @@
+import datetime
+import time
+
+
+SKIP_STALECHECK_FOR_METHODS = set([
+    'eth_getBlockByHash',
+    'eth_getBlockByNumber',
+    'eth_getBlockTransactionCountByHash',
+    'eth_getBlockTransactionCountByNumber',
+    'eth_getTransactionByBlockHashAndIndex',
+    'eth_getTransactionByBlockNumberAndIndex',
+    'eth_getUncleCountByBlockHash',
+    'eth_getUncleCountByBlockNumber',
+])
+
+
+def _isfresh(block, allowable_delay):
+    return block and time.time() - block.timestamp <= allowable_delay
+
+
+def make_stalecheck_middleware(**kwargs):
+    '''
+    Use to require that a function will run only of the blockchain is recently updated.
+
+    This middleware takes an argument, so unlike other middleware, you must make the middleware
+    with a method call.
+    For example: `make_stalecheck_middleware(hours=12)`
+
+    If the latest block in the chain is older than 12 hours ago in this example, then the
+    middleware will raise a StaleBlockchain exception.
+
+    `make_stalecheck_middleware(...)` takes the same keyword arguments as datetime.timedelta()
+    '''
+    allowable_delay = datetime.timedelta(**kwargs).total_seconds()
+
+    def stalecheck_middleware(make_request, web3):
+        def middleware(method, params):
+            if method not in SKIP_STALECHECK_FOR_METHODS:
+                last_block = web3.eth.getBlock('latest')
```

Comment:
So I think we can cache this.

* we only allow `D` seconds of *staleness*.
* Let `B` be the last block time we saw.

We should only need to fetch the latest block *if*.

* This is our first call and `B` is not populated.
* `time.time() - B['timestamp'] > D`

The idea is that as long as the time difference between now and the last block we saw is under the allowed delta, then even if the chain froze after retrieving the last block, we are still safe.  Only once the last block has expired (the timestamp is now beyond the current allowable delta) should we need to fetch the new latest block.

Also, we can be clever and try using the result of the RPC call to re-populate the latest block anytime we see that `method == "eth_getBlockByNumber" and params == ["latest"]`.

Reply:
Absolutely. I was thinking the same thing, but hadn't gotten around to it. I'll add it.

I might punt that last bit until the more general caching middleware. Trying to be clever usually gets me in trouble. :)

Reply:
Ugh, calling `web3.setProvider()` should invalidate the cache. I'm working on how to do that cleanly.

Reply:
I indexed into the cache with the providers found in web3. As long as people don't mutate the provider, we will be good. :crossed_fingers:

Outside of this PR, I'm tempted to say that we should remove `setProviders()` and force people to always start a new `Web3()`. Anything stateful is going to have to address this same problem.